File: pytrace/src/image.py
from color import Color
from PIL import Image as PILImage

class Image:
    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height
        self.pixels = [Color(0, 0, 0) for _ in range(width * height)]

    def set_pixel(self, x: int, y: int, color: Color):
        self.pixels[y * self.width + x] = color

    def save_as_ppm(self, file_path:str):
        with open(file_path, 'w') as f:
            self.write(f)

    def write(self, f):
        # header
        f.write(f"P3 {self.width} {self.height}" + EOL)
        f.write(f"{COLORS}" + EOL)

        def clamp(v, low, high):
            return max(min(v, high), low)

        def to_byte(c):
            return clamp(int(COLORS * c), 0, 255)

        for i, c in enumerate(self.pixels):
            f.write(f"{to_byte(c.x)} {to_byte(c.y)} {to_byte(c.z)} ")
            if (i + 1) % self.width == 0:
                f.write(EOL)
            else:
                f.write("\t")

    def save_as_png(self, file_path:str):
        img = PILImage.new("RGB", (self.width, self.height))
        for y in range(self.height):
            for x in range(self.width):
                c = self.pixels[y * self.width + x]
                img.putpixel((x, y), ((int(255 * c.x), int(255 * c.y), int(255 * c.z))))
        img.save(file_path)

EOL = "\n"
COLORS = 255

File: rust-tracer/src/image.rs
use crate::Color;
use image::{ImageBuffer, Rgb};

pub struct Image {
    pub width: u32,
    pub height: u32,
    pub pixels: Vec<Color>,
}

impl Image {
    pub fn new(width: u32, height: u32) -> Self {
        Image {
            width,
            height,
            pixels: vec![Color::new(0.0, 0.0, 0.0); (width * height) as usize],
        }
    }

    pub fn set_pixel(&mut self, x: u32, y: u32, color: Color) {
        self.pixels[(y * self.width + x) as usize] = color;
    }

    pub fn save_as_png(&self, file_path: &str) -> Result<(), image::ImageError> {
        let mut img = ImageBuffer::new(self.width, self.height);

        for (i, color) in self.pixels.iter().enumerate() {
            let x = (i as u32) % self.width;
            let y = (i as u32) / self.width;
            img.put_pixel(
                x, 
                y, 
                Rgb([
                    (color.vector.x * 255.0) as u8,
                    (color.vector.y * 255.0) as u8,
                    (color.vector.z * 255.0) as u8,
                ])
            );
        }

        img.save(file_path)?;
        Ok(())
    }
} 